

# 快速失败

https://blog.csdn.net/chewbee/article/details/78314661

## 1、引入：

​	通过迭代器访问非并发安全的集合类时，如果迭代过程，通过集合类自己的方法修改集合，会触发“快速失败”，抛出  “并发修改异常”  。

```java
public static void main(String[] args)  throws IOException {
        HashMap<String, String> map = new HashMap();
        map.put("0", "1");
        map.put("1", "x");
        map.put("2", "x");
        Iterator<String> i = map.keySet().iterator();
        while (i.hasNext()) {
            if (map.get(i.next()).equals("x")) {
                //此操作抛异常
                map.put("ok", "true");
                //下面操作不会出异常
                i.remove();
            }
        }
    }
```

## 2、原因：

```
以HashMap为例，其实现的Map接口包含与Iterator。
HashMap这些集合本身含有两个实例 ： int modCount 和 int excpectedModCount ;
通过迭代器迭代遍历时，刚开始会赋值， 让 modCount = expectedModCount ;
集合类通过自己方法修改集合时会修改 modCount ， 但是不会修改后者；而迭代器迭代修改则会；
所以，迭代器迭代之前会先判断两者是否相等，相等才迭代，不等则抛出并发修改异常。为的就是并发不安全提醒。


注意，迭代器的快速失败行为不能得到保证，一般来说，存在不同步的并发修改时，不能作出任何坚决的保证。快速失败迭代器尽最大努力抛出 。
因为HashMap本身不是并发安全的，并发修改可能存在 modCount = expectedModCount 但是过了这个判断后续的操作就存在并发错误。尽最大努力抛出！！！
```



# 安全失败

​	把上面的例子修改为ConcurrentHashMap就不会报错“ConcurrentModificationException” 了，虽然底层也是会出现不相等的情况。但是会撤回重新修改。为什么，因为，CAS+synchronized 。

ConcurrentHashMap 1.7 是使用 CAS+synchronized  ；而1.8开始完全采用 CAS。

失败重试，安全失败就是ConcurrentHashMap弱一致性的体现，虽然并发过程修改失败了出现错误，但不会抛出，而是回退重新修改，直到成功。



